src/hotspot/share/jfr/periodic/sampling/jfrThreadSampler.cpp

```diff
@@ -403,12 +407,12 @@ bool JfrThreadSampleClosure::do_sample_thread(JavaThread* thread, JfrStackFrame*
     SystemMemoryBarrier::emit();
   }
   if (JAVA_SAMPLE == type) {
-    if (thread_state_in_java(thread)) {
+    if (thread_state_in_java(thread) && !is_vthread_in_transition(thread)) {
       ret = sample_thread_in_java(thread, frames, max_frames);
     }
   } else {
     assert(NATIVE_SAMPLE == type, "invariant");
-    if (thread_state_in_native(thread)) {
+    if (thread_state_in_native(thread) && !is_vthread_in_transition(thread)) {
```

Comment:
Is this possible? I assume the thread is in _thread_in_Java during the transition?

Reply:
There are some native methods that we execute during mount/unmount transitions. From what I see they all seem to be defined as `@IntrinsicCandidate`, but if for some reason we don't execute the intrinsic version (interp only mode for instance) then we would call a normal native method.

Reply:
Just to ad that  Patricio suggested today to run the stress tests with -Xint and that does lead to triggered the assert quickly when the thread is sampled in native. There are several native methods that are @IntrinsicCandidate that are invoked after the thread identity has changed to the virtual thread and before the continuation is mounted. Probably less likely on the unmount as the only native method is the one that reverts the thread identity.

Reply:
Ok. If the thread can run native code as part of the mount / unmount, and the sampler can see the intermediary state there too, the check is needed.

Reply:
Thanks. One other thing that I see when more testing with generational ZGC is that ZPageAllocator::alloc_page will record an event and this can happen during a mount transition. So I think JfrStackTrace::record also needs to test if the current thread is a virtual thread without a continuation mounted.

Reply:
That is not good. Such a check would be attributed to all event sites, making them all slower.

Reply:
Markus has suggested, and provided the change, to handle the synchronous case in JfrStackTrace. The testing is looking good with the combined changes.